src/EditorFeatures/Core/SpellCheck/RoslynSpellCheckFixerProvider.cs

```diff
@@ -0,0 +1,110 @@
+            if (success)
+                return;
+
+            // failure path, just apply the text change directly.
+            await ApplySimpleChangeAsync(span, replacement, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async Task ApplySimpleChangeAsync(SnapshotSpan span, string replacement, CancellationToken cancellationToken)
+        {
+            await _threadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var buffer = span.Snapshot.TextBuffer;
+            var edit = buffer.CreateEdit();
+
+            edit.Replace(span.TranslateTo(buffer.CurrentSnapshot, SpanTrackingMode.EdgeInclusive), replacement);
+            edit.Apply();
+        }
+
+        private async Task<bool> TryRenameWordAsync(SnapshotSpan span, string replacement, CancellationToken cancellationToken)
+        {
+            // See if we can map this to a roslyn document.
+            var snapshot = span.Snapshot;
+            var document = snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document is null)
+                return false;
+
+            // If so, see if the language supports smart rename capabilities.
+            var renameService = document.GetLanguageService<IEditorInlineRenameService>();
+            if (renameService is null)
+                return false;
+
+            // Attempt to figure out what the language would rename here given the position of the misspelled word in
+            // the full token.
+            var info = await renameService.GetRenameInfoAsync(document, span.Span.Start, cancellationToken).ConfigureAwait(false);
+            if (!info.CanRename)
+                return false;
+
+            // The subspan we're being asked to rename better fall entirely within the span of the token we're renaming.
+            var fullTokenSpan = info.TriggerSpan;
+            var subSpanBeingRenamed = span.Span.ToTextSpan();
+            if (!fullTokenSpan.Contains(subSpanBeingRenamed))
+                return false;
+
+            // The language gave us the full token span, and we know the subportion to rename, so use all of that to
+            // form the final full token text to rename to.
+            var fullTokenReplacementText =
+                snapshot.GetSpanFromBounds(fullTokenSpan.Start, subSpanBeingRenamed.Start).GetText() +
+                replacement +
+                snapshot.GetSpanFromBounds(subSpanBeingRenamed.End, fullTokenSpan.End).GetText();
+
+            // Now attempt to call into the language to actually perform the rename.
+            var options = new SymbolRenameOptions();
+            var renameLocations = await info.FindRenameLocationsAsync(options, cancellationToken).ConfigureAwait(false);
+            var replacements = await renameLocations.GetReplacementsAsync(fullTokenReplacementText, options, cancellationToken).ConfigureAwait(false);
+            if (!replacements.ReplacementTextValid)
```

Comment:
no.  it's more 'false' if the name the user provided (or in this case, the editor) is not a legal name in teh language.  for example, if (for some reason) the editor said "rename `Wrd` to `W[]rd`" :)

Reply:
We definitely should not be providing anything like that, since these suggestion/replacements will be coming from the language dictionary in the first place. I have a bug to ensure this behavior though.

Reply:
@hjihong I'm not sure if you could actually prevent this though.   The words from teh dictionary may not be legal *language* identifiers (for any number of semantic resources).  So, ultimately, unless you have a way to check with us first on the viability of a dictionary-suggestion, this part may always be an issue :)